include/seqan3/core/algorithm/configuration.hpp

```diff
@@ -0,0 +1,1071 @@
+    //!}
+
+public:
+
+    /*!\brief Invokes the configuration specific functor to extend the seqan3::detail::configuration with the associated
+     *        config.
+     * \tparam configuration_t The configuration to be extended with the new configuration type. Must be of type
+     *                        seqan3::detail::configuration.
+     * \tparam args_t A template parameter pack with the arguments applied to the configuration functor call.
+     *
+     * \param configuration The configuration to be extended.
+     * \param args         The argument pack to be passed to the functor.
+     * \returns A new seqan3::detail::configuration with the extended configuration.
+     */
+    template <typename configuration_t,
+              typename ... args_t>
+    constexpr auto operator()(configuration_t && configuration,
+                              args_t && ... args) const
+        requires is_algorithm_configuration_v<remove_cvref_t<configuration_t>>
+    {
+        return static_cast<derived_fn_t const &>(*this).invoke(std::forward<configuration_t>(configuration),
+                                                               std::forward<args_t>(args)...);
+    }
+
+    /*!\brief Creates a proxy caching the arguments that should be applied when invoking the configuration specific
+     *        configuration.
+     * \tparam args_t A template parameter pack with the arguments applied to the configuration functor call.
+     *
+     * \param args A pack containing the arguments.
+     * \returns A proxy type used to defer configuration invocation.
+     */
+    template <typename ... args_t>
+    constexpr configuration_fn_proxy<derived_fn_t, args_t...> operator()(args_t && ... args) const
+    {
+        return {{std::forward<args_t>(args)...}};
+    }
+};
+
+/*!\name Pipe Interface
+ * \relates seqan3::detail::configuration_fn_base
+ * \{
+ */
+
+/*!\brief Combines a seqan3::detail::configuration with an configuration adaptor.
+ * \tparam configuration_t The configuration type to be extended. Must be of type seqan3::detail::configuration.
+ * \tparam fn_t            Either an instance of seqan3::detail::configuration_fn_proxy or
+ *                         seqan3::detail::configuration_fn_base
+ *
+ * \param[in] cfg The configuration to be extended.
+ * \param[in] fn  The functor implementation, invoking the associated configuration adaptor.
+ * \returns The result of invoking the configuration adaptor with the passed `cfg` object.
+ */
+template <typename configuration_t,
+          typename fn_t>
+constexpr auto operator|(configuration_t && cfg,
+                         fn_t && fn)
+//!\cond
+    requires is_algorithm_configuration_v<remove_cvref_t<configuration_t>> &&
+             (is_type_specialisation_of_v<remove_cvref_t<fn_t>, configuration_fn_proxy> ||
+              std::is_base_of_v<configuration_fn_base<remove_cvref_t<fn_t>>, remove_cvref_t<fn_t>>)
+//!\endcond
+{
+    return fn(std::forward<configuration_t>(cfg));
+}
```

Comment:
why not friend of the CRTP-base

Reply:
Told you. Constraint friend templates won’t work.